src/EventListener/ExceptionListener.php

```diff
@@ -0,0 +1,416 @@
+
+            return $response;
+        } catch (ResponseException $pageException) {
+            $processing = false;
+
+            return $pageException->getResponse();
+        } catch (NotFoundHttpException $pageException) {
+            $processing = false;
+
+            // We can safely assume that when we get an not found exception when trying to render a not found
+            // page, that we have to render the internal page.
+            return null;
+        } catch (\Exception $pageException) {
+            $processing = false;
+            // Throwing the page exception here is intentional as we have some error handler problems!
+            // We should never end up here.
+            // Fixing error handling code paths is more important than the original exception.
+            throw $pageException;
+        }
+     }
+
+    /**
+     * Check if the exception is understood by us.
+     *
+     * @param \Exception $exception The exception to walk on.
+     *
+     * @param Request    $request   The HTTP request.
+     *
+     * @return null|Response
+     */
+    private function checkHttpException($exception, $request)
+    {
+        if (!$exception instanceof HttpExceptionInterface) {
+            return null;
+        }
+
+        $response = $this->tryToRenderContao404($exception);
+        if ($response) {
+            return $response;
+        }
+
+        // Determine if the class or any of the parents is known by us.
+        $candidates = array_intersect(
+            array_merge([get_class($exception)], class_parents($exception)),
+            array_keys(self::$exceptionTemplates)
+        );
+        if (!empty($candidates)) {
+            $template = self::$exceptionTemplates[array_shift($candidates)];
+            return $this->createTemplateResponseFromException($template, $exception, $request);
+        }
+
+        // Unknown HttpExceptionInterface implementing exception, no way to handle.
+        return null;
+    }
+
+    /**
+     * Try to create a response for the given template.
+     *
+     * @param string                 $template  The name of the template to render.
+     *
+     * @param HttpExceptionInterface $exception The exception to render.
+     *
+     * @param Request                $request   The HTTP request.
+     *
+     * @return Response
+     */
+    private function createTemplateResponseFromException($template, HttpExceptionInterface $exception, $request)
+    {
+        $response = $this->renderErrorTemplate($template, $exception->getStatusCode(), $request);
+        $response->headers->add($exception->getHeaders());
+
+        return $response;
+    }
+
+    /**
+     * Load the language strings
+     *
+     * @return array|null
+     *
+     * @internal
+     *
+     * FIXME: replace this with a real - non legacy - language string loader.
+     */
+    protected function loadLanguageStrings()
+    {
+        System::loadLanguageFile('exception');
```

Comment:
This can only be called if the legacy framework is booted. Are we certain it is?